CRM/Core/BAO/File.php

```diff
@@ -261,51 +261,32 @@ public static function deleteEntityFile($entityTable, $entityID, $fileTypeID = N
     [$sql, $params] = self::sql($entityTable, $entityID, $fileTypeID, $fileID);
     $dao = CRM_Core_DAO::executeQuery($sql, $params);
 
-    $cfIDs = [];
-    $cefIDs = [];
+    $cfIDs = $cefIDs = [];
     while ($dao->fetch()) {
       $cfIDs[$dao->cfID] = $dao->uri;
-      $cefIDs[] = $dao->cefID;
+      $cefIDs[$dao->cfID] = $dao->cefID;
     }
 
-    // Delete tags from entity tag table.
     if (!empty($cfIDs)) {
-      $deleteFiles = [];
-      foreach ($cfIDs as $fId => $fUri) {
+      foreach ($cfIDs as $fileID => $fUri) {
         $tagParams = [
           'entity_table' => 'civicrm_file',
-          'entity_id' => $fId,
+          'entity_id' => $fileID,
         ];
+        // Delete tags from entity tag table.
         CRM_Core_BAO_EntityTag::del($tagParams);
-      }
-    }
-
-    // Delete entries from entity file table.
-    if (!empty($cefIDs)) {
-      $cefIDs = implode(',', $cefIDs);
-      $sql = "DELETE FROM civicrm_entity_file where id IN ( $cefIDs )";
-      CRM_Core_DAO::executeQuery($sql);
-      $isDeleted = TRUE;
-    }
 
-    if (!empty($cfIDs)) {
-      $deleteFiles = [];
-      foreach ($cfIDs as $fId => $fUri) {
+        // sequentially deletes EntityFile entry and then deletes File record
+        CRM_Core_DAO_EntityFile::deleteRecord(['id' => $cefIDs[$fileID]]);
         // Delete file only if there are no longer any entities using this file.
-        if (!CRM_Core_DAO::getFieldValue('CRM_Core_DAO_EntityFile', $fId, 'id', 'file_id')) {
+        if (!CRM_Core_DAO::getFieldValue('CRM_Core_DAO_EntityFile', $fileID, 'id', 'file_id')) {
```

Comment:
This is pre-existing but because this function is cached, if you e.g. delete two activities at the same time, e.g. from activity search, it deletes the entityfiles, but not the actual file. I suppose that's only a diskspace/integrity issue, and not loss of data so, again, 🤷‍♂️ .